truffle/src/main/java/org/jruby/truffle/nodes/ext/BigDecimalNodes.java

```diff
@@ -178,175 +217,327 @@ public static boolean isNan(RubyBasicObject value) {
             return getBigDecimalType(value) == Type.NAN;
         }
 
+        private void setupCreateBigDecimal() {
+            if (createBigDecimal == null)
+                createBigDecimal = insert(CreateBigDecimalNodeFactory.create(getContext(), getSourceSection(), null, null));
+        }
+
+        protected RubyBasicObject createBigDecimal(VirtualFrame frame, Object value) {
+            setupCreateBigDecimal();
+            return createBigDecimal.executeCreate(frame, value);
+        }
+
+        protected RubyBasicObject createBigDecimal(VirtualFrame frame, Object value, RubyBasicObject self) {
+            setupCreateBigDecimal();
+            return createBigDecimal.executeCreate(frame, value, self);
+        }
+
+        private void setupLimitCall() {
+            if (limitCall == null)
+                limitCall = insert(DispatchHeadNodeFactory.createMethodCall(getContext()));
```

Comment:
But like exceptions that doesn't cover the `DispatchHeadNodeFactory.createMethodCall(getContext())` part, or the method call to `insert`. We want to transfer immediately when we find we have to and not have any of the slow path code in the compiled part.

We've been using just `transfer` in the caller. The `insert` does the `invalidate`.